src/main/java/seedu/address/ui/UiManager.java

```diff
@@ -125,7 +128,42 @@ public void setGuiSettings(GuiSettings guiSettings) {
         logic.setGuiSettings(guiSettings);
     };
 
+    public MainWindow getMainWindow() {
+        return this.mainWindow;
+    }
+
+    /**
+     * Passes the string to logic to execute. Gets the command result back from logic, along with any
+     * uiConsumer. If there is one, apply it to the ui
+     * @param commandText String entered by the user
+     * @return  result of the command of class Command result
+     * @throws CommandException
+     * @throws ParseException
+     */
     public CommandResult execute(String commandText) throws CommandException, ParseException {
-        return logic.execute(commandText);
+        CommandResult commandResult = logic.execute(commandText);
+        Consumer<Ui> uiConsumer = commandResult.getUiAction();
+        if (!uiConsumer.equals(null)) {
+            uiConsumer.accept(this);
+        }
+        return commandResult;
     };
+
+    //For testing purposes. See UiManager Test for more info
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
```

Comment:
Nice, thanks for preparing this for the test!

```suggestion
        if (o == null || o instanceof UiManager) {
```